### PR TITLE
Feature/kong upgrade

### PR DESCRIPTION
**Pull request by tingyuz**

### CODE REVIEW

**Comment by anonymouse64** [comment]
@tingyuz could you add a description here of the changes you made here to handle the kong bug with using the `Host` header?

**Comment by tingyuz** [comment]
@jpwhitemn 
1. This is not real token file,  it is just serving as an example. The real token file will be created during the initialization in the shared volume.
2. Let me switch that version - I was using the old one for testing and should have 1.0.3 in the Edinburgh release.

**Comment by tingyuz** [comment]
@anonymouse64 Yes I removed the host header totally as I think the bug from kong 1.0.3 is the combination of host header and the service/route/ACL. I updated the readme file to show the sample request from the client that host header is no longer needed.

**Comment by anonymouse64** [comment]
@tingyuz I won't be able to review this today (hopefully should be able to get to this week), but please don't merge until I've had a chance to review this. Thanks!

**Comment by tingyuz** [comment]
@anonymouse64  I made some changes regarding the OAuth2 configuration in the toml file, mainly reduced the configurable items to simplify the configuration. And I am applying global authorization and ACL for all the microservices so that one token created in one service can be applied for all the EdgeX services VS. one token for each service in the previous version.

File: core/init.go [hunk 1]
```diff
@@ -146,76 +163,108 @@ func initOauth2ForService(config *tomlConfig, url string, c *http.Client) {
...
+		EnableGlobalCredentials: "true",
 	}
 
-	req, err := sling.New().Base(url).Post(PluginsPath).BodyForm(jwtParams).Request()
+	req, err := sling.New().Base(url + "services/" + service + "/").Post(PluginsPath).BodyForm(oauth2Params).Request()
 	resp, err := c.Do(req)
 	if err != nil {
-		s := fmt.Sprintf("Failed to set up jwt authentication.")
+		s := fmt.Sprintf("Failed to set up oauth2 authentication with error %s.", err.Error())
+		defer resp.Body.Close()
```
Comment (anonymouse64) [inline core/init.go]: if the error was not nil here we don't need to close the body - it shouldn't have been allocated.

Instead put the `defer resp.Body.Close()` in the else clause here

File: core/init.go [hunk 2]
```diff
@@ -146,76 +163,108 @@ func initOauth2ForService(config *tomlConfig, url string, c *http.Client) {
...
 		lc.Error(s)
 	} else {
 		if resp.StatusCode == 200 || resp.StatusCode == 201 || resp.StatusCode == 409 {
```
Comment (anonymouse64) [inline core/init.go]: same comment as above about using numeric http status codes

File: core/init.go [hunk 3]
```diff
@@ -146,76 +163,108 @@ func initOauth2ForService(config *tomlConfig, url string, c *http.Client) {
...
+	jwtParams := &KongJWTPlugin{
+		Name: config.KongAuth.Name,
+	}
+
+	req, err := sling.New().Base(url).Post(PluginsPath).BodyForm(jwtParams).Request()
 	resp, err := c.Do(req)
 	if err != nil {
-		s := fmt.Sprintf("Failed to set up routes for %s with error %s.", name, err.Error())
+		s := fmt.Sprintf("Failed to set up jwt authentication.")
+		defer resp.Body.Close()
```
Comment (anonymouse64) [inline core/init.go]: same comment as above about closing the body in the else clause here

File: core/init.go [hunk 4]
```diff
@@ -146,76 +163,108 @@ func initOauth2ForService(config *tomlConfig, url string, c *http.Client) {
...
-		Host:     config.KongURL.Server,
-		Port:     config.KongURL.AdminPort,
-		Protocol: "http",
+func initJWTAuthForRoute(config *tomlConfig, url string, c *http.Client, service string) {
+	jwtParams := &KongJWTPlugin{
+		Name: config.KongAuth.Name,
 	}
-	req, err := sling.New().Base(url).Post(ServicesPath).BodyForm(adminServiceParams).Request()
+
+	req, err := sling.New().Base(url + "routes/" + service + "/").Post(PluginsPath).BodyForm(jwtParams).Request()
```
Comment (anonymouse64) [inline core/init.go]: same comment about using `filepath.Join` as above

File: core/init.go [hunk 5]
```diff
@@ -146,76 +163,108 @@ func initOauth2ForService(config *tomlConfig, url string, c *http.Client) {
...
 	resp, err := c.Do(req)
 	if err != nil {
-		s := fmt.Sprintf("Failed to set up service for admin loopback with error %s.", err.Error())
+		s := fmt.Sprintf("Failed to set up jwt authentication.")
+		defer resp.Body.Close()
```
Comment (anonymouse64) [inline core/init.go]: same comment as above for closing body in else clause

File: core/init.go [hunk 6]
```diff
@@ -146,76 +163,108 @@ func initOauth2ForService(config *tomlConfig, url string, c *http.Client) {
...
 		}
 	}
+}
 
-	adminRouteParams := &KongRoute{Paths: []string{"/admin"}}
-	adminRoutePath := fmt.Sprintf("%sadmin/routes", ServicesPath)
-	req, err = sling.New().Base(url).Post(adminRoutePath).BodyJSON(adminRouteParams).Request()
-	resp, err = c.Do(req)
+func initKongRoutes(url string, c *http.Client, r *KongRoute, name string) {
+	routesubpath := "services/" + name + "/routes"
```
Comment (anonymouse64) [inline core/init.go]: use filepath.Join here too

File: core/init.go [hunk 7]
```diff
@@ -146,76 +163,108 @@ func initOauth2ForService(config *tomlConfig, url string, c *http.Client) {
...
+	lc.Info(routesubpath)
```
Comment (anonymouse64) [inline core/init.go]: this message should be more descriptive about what it is printing off to the logs

File: core/init.go [hunk 8]
```diff
@@ -146,76 +163,108 @@ func initOauth2ForService(config *tomlConfig, url string, c *http.Client) {
...
+	req, err := sling.New().Base(url).Post(routesubpath).BodyJSON(r).Request()
+	resp, err := c.Do(req)
 	if err != nil {
-		lc.Error("Failed to set up admin service route.")
+		s := fmt.Sprintf("Failed to set up routes for %s with error %s.", name, err.Error())
+		defer resp.Body.Close()
```
Comment (anonymouse64) [inline core/init.go]: same comment as above about closing body in the else clause

File: core/init.go [hunk 9]
```diff
@@ -146,76 +163,108 @@ func initOauth2ForService(config *tomlConfig, url string, c *http.Client) {
...
+		lc.Error(s)
 	} else {
 		if resp.StatusCode == 200 || resp.StatusCode == 201 || resp.StatusCode == 409 {
```
Comment (anonymouse64) [inline core/init.go]: same comment as above about numeric http status codes

File: core/token.go [hunk 10]
```diff
@@ -40,3 +45,19 @@ func getSecret(filename string) (string, error) {
...
+		User:  u,
+		Token: t,
+	}
+
+	jdata, err := json.MarshalIndent(data, "", " ")
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(filename, jdata, 0644)
```
Comment (anonymouse64) [inline core/token.go]: should this token be created by default with `0600` mode? As it is currently it is world-readable, which for an access token seems undesirable

**Comment by tingyuz** [comment]
@anonymouse64 update most of the part in the codebase based on your suggestion. i will have another PR to address the httpstatus code issue you mentioned as it spreads out in the code a couple of places, it might be better to have another PR for it.

**Comment by anonymouse64** [comment]
@tingyuz thanks, I'll take a look soon. And also that's fine to have another PR to address the http status code issue, just file an issue about it so we don't forget

**Comment by tingyuz** [comment]
Open issu #57 to track the change of status code per suggestion from @anonymouse64.

File: core/init.go [hunk 11]
```diff
@@ -139,83 +152,115 @@ func initOauth2ForService(config *tomlConfig, url string, c *http.Client) {
...
 		} else {
 			s := fmt.Sprintf("Failed to set up oauth2 authentication with errorcode %d.", resp.StatusCode)
 			lc.Error(s)
 		}
 	}
+}
 
+func initAuthmethod(config *tomlConfig, url string, c *http.Client) {
+	lc.Info(fmt.Sprintf("selected auth method as %s.", config.KongAuth.Name))
+	if config.KongAuth.Name == "jwt" {
```
Comment (anonymouse64) [inline core/init.go]: a `switch` here might be nice if we anticipate adding support for other auth methods

Comment (tingyuz) [reply core/init.go]: OK I will get it into the place for the next PR.

File: docker-compose-edinburgh-1.0.0.yml [hunk 12]
```diff
@@ -127,18 +127,18 @@ services:
         - 'POSTGRES_USER=kong'
 
   kong-migrations:
-    image: "kong:0.13.0"
+    image: "kong:1.0.3"
```
Comment (hutchic) [inline docker-compose-edinburgh-1.0.0.yml]: This image isn't arm compatible
```
ubuntu@ip-172-31-27-205:~$ uname -m
aarch64
ubuntu@ip-172-31-27-205:~$ docker run -it --rm kong:1.0.3 /bin/sh
Unable to find image 'kong:1.0.3' locally
1.0.3: Pulling from library/kong
docker: no matching manifest for unknown in the manifest list entries.
```

Comment (anonymouse64) [reply docker-compose-edinburgh-1.0.0.yml]: Currently, the project doesn't use multi-arch manifests and publishes different image names for arm than for x86

Comment (hutchic) [reply docker-compose-edinburgh-1.0.0.yml]: @anonymouse64 is this reply about EdgeX not using multi-arch or Kong?

Comment (anonymouse64) [reply docker-compose-edinburgh-1.0.0.yml]: sorry EdgeX not using multi-arch, so the docker-compose file you're referencing is meant to be x86 specific and there will be a separate one for armhf or arm64

Comment (anonymouse64) [reply docker-compose-edinburgh-1.0.0.yml]: See https://github.com/edgexfoundry/developer-scripts/issues/104 also

Comment (hutchic) [reply docker-compose-edinburgh-1.0.0.yml]: Ah perfect I was concerned EdgeX was going to try to run ☝️ on non x86 as is which I predict sadness :)

**Pull request merged**